safehttp/plugins/cors/cors.go

```diff
@@ -0,0 +1,214 @@
+}
+
+// SetAllowedHeaders sets which headers should be allowed in the Access-Control-Allow-Headers
+// header. The headers are first canonicalized using textproto.CanonicalMIMEHeaderKey.
+func (it *Interceptor) SetAllowedHeaders(headers ...string) {
+	it.allowedHeaders = map[string]bool{}
+	for _, h := range headers {
+		if h == "*" {
+			continue
+		}
+		it.allowedHeaders[textproto.CanonicalMIMEHeaderKey(h)] = true
+	}
+}
+
+// Before handles the incoming request and sets headers or responds with an error
+// accordingly.
+func (it *Interceptor) Before(w safehttp.ResponseWriter, r *safehttp.IncomingRequest) safehttp.Result {
+	origin := r.Header.Get("Origin")
+	if origin != "" && !it.AllowedOrigins[origin] {
+		return w.ClientError(safehttp.StatusForbidden)
+	}
+	h := w.Header()
+	allowOrigin, err := h.Claim("Access-Control-Allow-Origin")
+	if err != nil {
+		return w.ServerError(safehttp.StatusInternalServerError)
+	}
+	vary, err := h.Claim("Vary")
+	if err != nil {
+		return w.ServerError(safehttp.StatusInternalServerError)
+	}
+
+	allowCredentials, err := h.Claim("Access-Control-Allow-Credentials")
+	if err != nil {
+		return w.ServerError(safehttp.StatusInternalServerError)
+	}
+
+	var status safehttp.StatusCode
+	if r.Method() == safehttp.MethodOptions {
+		status = it.preflight(w, r)
+	} else {
+		status = it.request(w, r)
+	}
+
+	if status != 0 && status != safehttp.StatusNoContent {
+		if 400 <= status && status < 500 {
+			return w.ClientError(status)
+		} else {
+			return w.ServerError(status)
+		}
+	}
+
+	if origin != "" {
+		allowOrigin([]string{origin})
+		vary([]string{"Origin"})
+	}
+	if r.Header.Get("Cookie") != "" && it.AllowCredentials {
+		// TODO: handle other credentials than cookies:
+		// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Credentials
+		allowCredentials([]string{"true"})
+	}
+
+	if status == safehttp.StatusNoContent {
+		return w.NoContent()
+	}
+	return safehttp.Result{}
+}
+
+// preflight handles requests that have the method OPTIONS.
+func (it *Interceptor) preflight(w safehttp.ResponseWriter, r *safehttp.IncomingRequest) safehttp.StatusCode {
+	rh := r.Header
+	method := rh.Get("Access-Control-Request-Method")
+	if method == "" {
+		return safehttp.StatusForbidden
+	}
+	wh := w.Header()
+	allowMethods, err := wh.Claim("Access-Control-Allow-Methods")
+	if err != nil {
+		return safehttp.StatusInternalServerError
+	}
+
+	headers := rh.Get("Access-Control-Request-Headers")
+	if headers != "" {
+		for _, h := range strings.Split(headers, ", ") {
+			h = textproto.CanonicalMIMEHeaderKey(h)
+			if !it.allowedHeaders[h] && h != requiredHeader {
+				return safehttp.StatusForbidden
+			}
+		}
+	}
+	allowHeaders, err := wh.Claim("Access-Control-Allow-Headers")
+	if err != nil {
+		return safehttp.StatusInternalServerError
+	}
+
+	maxAge, err := wh.Claim("Access-Control-Max-Age")
+	if err != nil {
+		return safehttp.StatusInternalServerError
+	}
+
+	allowMethods([]string{method})
```

Comment:
Where do you check if the method is allowed? We still want to block HEAD requests.

Reply:
Can you send HEAD requests with `X-Cors: 1` and a non-standard Content-Type cross site without triggering a preflight? It's for this reason that we allowed GET and POST, so why ban HEAD?

Reply:
We have observed over the years that head requests could be crafted via browser bugs.

Also no well intended application I have ever seen uses HEAD requests.